esmvalcore/esgf/_download.py

```diff
@@ -0,0 +1,450 @@
+                if hasher is not None:
+                    hasher.update(chunk)
+                file.write(chunk)
+
+        duration = datetime.datetime.now() - start_time
+
+        if hasher is None:
+            logger.warning(
+                "No checksum available, unable to check data"
+                " integrity for %s, ", url)
+        else:
+            local_checksum = hasher.hexdigest()
+            if local_checksum != checksum:
+                raise DownloadError(
+                    f"Wrong {checksum_type} checksum for file {tmp_file},"
+                    f" downloaded from {url}: expected {checksum}, but got"
+                    f" {local_checksum}. Try downloading the file again.")
+
+        shutil.move(tmp_file, local_file)
+        log_speed(url, self.size, duration.total_seconds())
+        logger.info("Downloaded %s (%.0f MB) in %s (%.1f MB/s) from %s",
+                    local_file, self.size / 2**20, duration,
+                    self.size / 2**20 / duration.total_seconds(),
+                    urlparse(url).hostname)
+
+
+def get_download_message(files):
+    """Create a log message describing what will be downloaded."""
+    megabyte = 2**20
+    gigabyte = 2**30
+    total_size = 0
+    lines = []
+    for file in files:
+        total_size += file.size
+        lines.append(f"{file.size / megabyte:.0f} MB" "\t" f"{file}")
+    if total_size:
+        lines.insert(0, "Will download the following files:")
+    lines.insert(0, f"Will download {total_size / gigabyte:.1f} GB")
+    return "\n".join(lines)
+
+
+def download(files, dest_folder, n_jobs=4):
+    """Download multiple ESGFFiles in parallel.
+
+    Arguments
+    ---------
+    files: list of :obj:`ESGFFile`
+        The files to download.
+    dest_folder: Path
+        The destination folder.
+    n_jobs: int
+        The number of files to download in parallel.
+
+    Raises
+    ------
+    DownloadError:
+        Raised if one or more files failed to download.
+    """
+    files = sorted(files)
+    logger.info(get_download_message(files))
+    logger.info("Downloading..")
+
+    def _download(file: ESGFFile):
+        """Download file to dest_folder."""
+        file.download(dest_folder)
+
+    total_size = 0
+    start_time = datetime.datetime.now()
+
+    errors = []
+    random.shuffle(files)
+    with concurrent.futures.ThreadPoolExecutor(max_workers=n_jobs) as executor:
+        future_to_file = {
+            executor.submit(_download, file): file
+            for file in files
+        }
+
+        for future in concurrent.futures.as_completed(future_to_file):
+            file = future_to_file[future]
+            try:
+                future.result()
+            except DownloadError as error:
+                logger.error("Failed to download %s, error message %s", file,
+                             error)
+                errors.append(error)
+            else:
+                total_size += file.size
```

Comment:
A recursion limit seems unlikely here, because there is no recursion taking place. There is already a timeout on the downloads, currently set to time out if the hosts that you're downloading from does not respond 5 minutes:
https://github.com/ESMValGroup/ESMValCore/blob/72eee4d1f4ade5bdbe534af04326765a26b42240/esmvalcore/esgf/_download.py#L25-L26

Reply:
thanks man, I looked at the code again, the `ESGFFile` object looks fine and looks to play will with concurrent futures - btw you have no idea how many times you may hit the max recursion depth issue even if there is no obvious recursion in sight :grin: Cool about the timer, missed that one! But 5min is prob too long don't you think?